File: 23/pset23b.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <algorithm>
#include <iterator>
#include "Node.cpp"

using namespace std;

const int MOVES = 10000000;
const int CUPS = 1000000;

int main() {

    ifstream myfile("input.txt");
    std::map<int, Node*> nodes;
    std::string inputString;
    getline(myfile, inputString);

    // Initialise Linked List from input
    Node* ptrPrev = new Node();
    for (char iChar: inputString) {
        int i = iChar - '0';
        Node* ptrNew = new Node(i, NULL, NULL);
        if (ptrPrev->val) {
            ptrPrev->next = ptrNew;
        }
        nodes.insert(std::pair<int, Node*>(i, ptrNew));
        ptrPrev = ptrNew;
    }

    // Add the rest of the numbers from 1 - CUPS to the Linked List
    for (int i = inputString.length() + 1; i <= CUPS; ++i) {
        Node* ptrNew = new Node(i, NULL, NULL);
        if (ptrPrev->val) {
            ptrPrev->next = ptrNew;
        }
        nodes.insert(std::pair<int, Node*>(i, ptrNew));
        ptrPrev = ptrNew;
    }

    // Make Linked List circular
    Node* curr = nodes.at(inputString[0] - '0');
    ptrPrev->next = curr;

    // Start the game
    for (int i = 0; i < MOVES; ++i) {
        Node* cup1 = curr->next;
        Node* cup2 = cup1->next;
        Node* cup3 = cup2->next;
        curr->next = cup3->next;
        int cupValues[3] = {cup1->val, cup2->val, cup3->val};
        int destinationVal = curr->val - 1;
        if (destinationVal < 1) destinationVal = CUPS;
        while (std::find(cupValues, std::end(cupValues), destinationVal) != std::end(cupValues)) {
            --destinationVal;
            int destinationVal = curr->val - 1;
            if (destinationVal < 1) destinationVal = CUPS;
        }
        Node* destination = nodes.at(destinationVal);
        cup3->next = destination->next;
        destination->next = cup1;
        
        curr = curr->next;
    }

    // Find "1" cup
    while (curr->val != 1) {
        curr = curr->next;
    }
    
    cout << (long long) curr->next->val * (long long) curr->next->next->val << endl;

    return 0;
}

File: 23/pset23b.py
class Node:
    def __init__(self, val=None):
        self.val = val
        self.prev = None
        self.next = None

with open("input.txt", "r") as f:
    rawinput = f.read().splitlines()
input = []
for i in rawinput[0]:
    input.append(int(i))

nodes = {}

prev_node = None
for i in input:
    node = Node(i)
    nodes[i] = node
    if prev_node is not None:
        node.prev = prev_node
        prev_node.next = node
    prev_node = node

for i in range(len(input) + 1, 1000000 + 1):
    node = Node(i)
    nodes[i] = node
    if prev_node is not None:
        node.prev = prev_node
        prev_node.next = node
    prev_node = node

current = nodes[input[0]]
prev_node.next = current
current.prev = prev_node

for i in range(10000000):
    current_val = current.val

    cup_1 = current.next
    cup_2 = cup_1.next
    cup_3 = cup_2.next

    current.next = cup_3.next
    current.next.prev = current

    destination_val = current_val - 1
    if destination_val < 1:
        destination_val = 1000000
    while destination_val in (cup_1.val, cup_2.val, cup_3.val):
        destination_val -= 1
        if destination_val < 1:
            destination_val = 1000000
        
    
    destination = nodes[destination_val]
    cup_3.next = destination.next
    cup_3.next.prev = cup_3
    destination.next= cup_1
    cup_1.prev = destination

    current = current.next

    print(current.val)

while current.val != 1:
    current = current.next

print(current.next.val * current.next.next.val)